CONFETI-33 feat: [온보딩 뷰] 선택된 아티스트 다건 삭제 API 구현

jher235 opened the pull request:
## 🔊 Summaries
<!-- 본인이 한 작업을 요약해주세요. -->
- 선택된 아티스트 다건 삭제 API 구현
  - 현재 해당 api는 목록 삭제를 위해 존재함. 처음에는 HTTP DELETE를 사용하고자 했으나 DELETE 의 경우 RequestBody를 사용하는 것이 표준적이지 않음. 스프링은 지원하지만, 지원하지 않는 서버나 미들웨어 등이 존재하므로 PATCH 메서드를 통해 RequestBody로 목록을 받아서 처리하도록 구현함

## ✨ Notification 
<!-- 참고할 사항을 작성해주세요. -->

jher235 committed:
CONFETI-33 feat: 온보딩 중 좋아요 누른 아티스트 목록 수정을 위한 Dto 작성
- 현재는 수정 시 delete 만 가능하므로 필드로 deleteFavoriteArtistIds 만 존재함

jher235 committed:
CONFETI-33 feat: 온보딩 중 좋아요 누른 아티스트 목록 삭제 로직 구현

jher235 committed:
CONFETI-33 feat: 온보딩 중 좋아요 누른 아티스트 목록 삭제 presentation layer 구현
- 현재 해당 api는 목록 삭제를 위해 존재함. 처음에는 HTTP DELETE를 사용하고자 했으나 DELETE 의 경우 RequestBody를 사용하는 것이 표준적이지 않음. 스프링은 지원하지만, 지원하지 않는 서버나 미들웨어 등이 존재하므로 PATCH 메서드를 통해 RequestBody로 목록을 받아서 처리하도록 구현함

jher235 committed:
CONFETI-33 feat: patchFavoriteArtists 메서드 명세 작성

ch1hyun commented on src/main/java/org/sopt/confeti/api/user/controller/docs/UserOnboardControllerV4Docs.java:
```diff
@@ -28,4 +29,25 @@ ResponseEntity<BaseResponse<UserOnboardFavoriteArtistsResponse>> getFavoriteArti
+    @AuthErrorResponses
+    @CommonErrorResponses
+    ResponseEntity<BaseResponse<Void>> patchFavoriteArtists(
+        Long userIdm,
```
클로드가 말한대로 오타가 있어요

ch1hyun commented on src/main/java/org/sopt/confeti/api/user/facade/UserOnboardFacade.java:
```diff
@@ -153,6 +154,20 @@ public void cacheTopArtists(List<ConfetiArtist> topArtists) {
+
+        boolean isRemoved = favoriteArtistIds.removeAll(deleteFavoriteArtistIds);
+        if (!isRemoved) {
+            throw new NotFoundException(ErrorMessage.NOT_FOUND);
```
Fail silent 전략을 사용하지 않으신 이유가 있으신가요?
만약 삭제 실패 시 Not Found 에러를 응답한다면, 존재하지 않는 아이디를 삭제 시도했을 때 예외 처리를 하지 않으신거는 의도하신건지 궁금합니다.

jher235 replied:
삭제 자체가 이뤄지지 않으면 존재하지 않는 id 값을 삭제하려는 잘못된 요청임을 알리기 위해 예외를 발생시키는게 안전하고 생각했어요.. 다만 지금 구조는 다소 어중간한 것 같네요. 
처음에 개별 삭제로 로직을 구현했다가 다건 삭제로 바꾸면서 이 부분을 놓쳤네요..
존재하는 id인지를 일일이 확인한 뒤 삭제하는 것과 fail silent를 사용하는 것 중 어떤게 더 좋다고 보세요??

jher235 replied:
조금 더 생각해봤는데 fail silent로 구현할까요?? 존재하지 않는 id를 삭제하려해도 큰 위험 요인이 될 것 같진 않고, 네트워크 이슈 등으로 중복 요청이 들어와도 멱등성 보장이 될 것 같네요

ch1hyun replied:
좋은 생각인 것 같습니다!! 동의해요

jher235 committed:
CONFETI-33 chore: 오타수정

jher235 committed:
CONFETI-33 feat: 좋아요 누른 아티스트 목록 삭제 로직을 fail silent 하게 처리하도록 변경
- 기존 로직은 삭제를 완전히 실패한 경우에만 오류를 발생시키고 여러 id 중 하나만 존재해도 오류가 발생하지 않는 애매한 검증을 하고 있었음. 멱등성 보장 및 애매한 검증을 없애기 위해 fail silent하게 처리하도록 변경

jher235 committed:
CONFETI-33 feat: deleteFavoriteArtistIds 유효성 검사 추가
